lib/service.js

```diff
@@ -1,264 +1,244 @@
+      throw new Error('Unable to parse CFG');
+    }
 
-        if (serviceMethod === 'getMetaModel' || serviceMethod === 'selectModel') {
-            assert(ACTION_OR_TRIGGER, 'ELASTICIO_ACTION_OR_TRIGGER is not provided');
-        }
-        if (serviceMethod === 'selectModel') {
-            assert(GET_MODEL_METHOD, 'ELASTICIO_GET_MODEL_METHOD is not provided');
-        }
+    debug('Config: %j', cfg);
 
-        var cfg;
-        try {
-            cfg = JSON.parse(CFG);
-        } catch (e) {
-            throw new Error('Unable to parse CFG');
-        }
+    const params = {
+      triggerOrAction: ACTION_OR_TRIGGER,
+      getModelMethod: GET_MODEL_METHOD,
+    };
 
-        debug('Config: %j', cfg);
+    return [cfg, params];
+  }
+
+  function execService(cfg, params) {
+    debug('Init is complete. About to start execution.');
+
+    return compReader.init(COMPONENT_PATH).then(() => ALLOWED_METHODS[serviceMethod](cfg, params));
+  }
+
+  function sendResponse(responseBody) {
+    const opts = {
+      url: POST_RESULT_URL,
+      json: true,
+      forever: true,
+      headers: {
+        Connection: 'Keep-Alive',
+      },
+      rejectUnauthorized: false,
+      body: responseBody,
+      simple: false,
+      maxAttempts: parseInt(env.ELASTICIO_API_REQUEST_RETRY_ATTEMPTS, 10),
+      retryDelay: parseInt(env.ELASTICIO_API_REQUEST_RETRY_DELAY, 10),
+      retryStrategy: request.RetryStrategies.HTTPOrNetworkError,
+      fullResponse: true,
+    };
 
-        var params = {
-            triggerOrAction: ACTION_OR_TRIGGER,
-            getModelMethod: GET_MODEL_METHOD
-        };
+    debug('About to send response back to the API');
+
+    return request.post(opts)
+      .then((response) => {
+        // eslint-disable-next-line eqeqeq
+        if (response.statusCode != '200') {
+          debug('Unable to reach API :(');
+          const error = new Error(util.format(
+            'Failed to POST data to %s (%s, %s)',
+            POST_RESULT_URL, response.statusCode, response.body,
+          ));
+          error.sendable = false;
+          throw error;
+        }
+      })
+      .then(() => responseBody);
+  }
+
+  function onSuccess(data) {
+    return sendResponse({ status: 'success', data });
+  }
+
+  function onError(err) {
+    if (err.sendable === false) {
+      throw new Error(err.message);
+    }
+    const errorData = {
+      message: err.message,
+    };
+    log.error(err, err.stack);
+    return sendResponse({ status: 'error', data: errorData });
+  }
+
+  return Q.fcall(init)
+    .spread(execService)
+    .then(onSuccess)
+    .catch(onError);
+
+  async function verifyCredentials(cfg) {
+    const verify = await compReader.loadVerifyCredentials();
+
+    const caller = new EventEmitter();
+    caller.logger = new ComponentLogger();
+
+    let response;
+    let promiseResponse;
+
+    try {
+      response = await verify.call(caller, cfg, (e, verification) => {
+        if (e) {
+          throw e;
```

Comment:
this won't work correctly => it will generate unhandled rejection/exception
example: let verify function look like this
```
function verify(cfg, cb)  {
    setTimeout(() => cb(new Error('smth bad happened'));
}
```